src/django_oapif/renderers.py

```diff
@@ -0,0 +1,62 @@
+import io
+from typing import Any, Generator, OrderedDict
+
+import fiona
+import orjson
+import ujson
+from django.conf import settings
+from fiona.crs import CRS
+from json_stream_generator import json_generator
+from rest_framework import renderers
+
+
+class FGBRenderer(renderers.BaseRenderer):
+    format = "fgb"
+    media_type = "application/event-stream"
+
+    def render(self, data: OrderedDict, accepted_media_type=None, renderer_context=None) -> io.BytesIO:
+        """Renders data as a flatgeobuf binary stream"""
+        assert renderer_context
+        features_data = data["features"] if "features" in data else data["results"]["features"]
+        features = (fiona.Feature.from_dict(obj) for obj in features_data)
+        buffer_wrapper = io.BytesIO()
+        with fiona.open(
+            buffer_wrapper,
+            mode="w",
+            driver="FlatGeobuf",
+            schema=renderer_context["schema"],
+            crs=CRS.from_epsg(settings.GEOMETRY_SRID),
+        ) as fh:
+            for feature in features:
+                fh.write(feature)
+
+        buffer_wrapper.seek(0)
+        return buffer_wrapper
+
+
+class JSONStreamingRenderer(renderers.BaseRenderer):
+    format = "json"
+    media_type = "application/x-ndjson"
+
+    def render(self, data: OrderedDict, accepted_media_type=None, renderer_context=None) -> Generator[Any, None, None]:
+        """Renders JSON encoded stream."""
+        features_data = data["features"] if "features" in data else data["results"]["features"]
+        return json_generator(feature for feature in features_data)
```

Comment:
iterator object so that the response's rendered content can be streamed from the renderer. Using the default JSON renderer and / or the naked StreamingHttps response (naked as in: "without passing this constructor an iterator object") doesn't provide actual streaming.

Reply:
Just in case this PR remains open (or the other branch).

What is the type of `data["results"]["features"]`? If it is a list, wouldn't it be great to try to preserve it as a db cursor until this stage so we really experience the benefit of the streaming.

Otherwise if it is a list, we iterate over multiple times.

Also, I am afraid the datastructure returned by this response is just an array of features, while the default json renderer returns an object (as it is in the specs?).

Reply:
The `data` argument of this function is the serialized data, i.e. in Djano parlance, the result of applying the serializer to the QuerySet. By definition, render() functions always apply after serialize() functions. So whatever the type of datæ["results"]["features"] is, it is a serialized type, constructing after closing the cursor.

As for `json_generator`: https://github.com/wlatanowicz/json-stream-generator/blob/0a3aa95acca188f2587da3b6f5c4915d49b0ee39/json_stream_generator/__init__.py#L19

Reply:
> The data argument of this function is the serialized data. So whatever the type of data["results"]["features"] is, it is a serialized type, constructed after closing/consuming the cursor. I think there's an advantage in this; if you let the user consuming the response get a hold of the database cursor, you are making the cursor a hostage of the connection. This requires some extra complexity (i.e. managing different sorts of exceptions differently depending on their nature) which I didn't think was worth it for the sake of this prototype.

Agree this brings additional complexity we might not need for now. Just worth mentioning (and documenting?) one of the main benefits of extremely big responses is not to fit everything in the memory.

> What is "this response"? json_generator? It can be seen here [wlatanowicz/json-stream-generator@0a3aa95/json_stream_generator/__init__.py#L19](https://github.com/wlatanowicz/json-stream-generator/blob/0a3aa95acca188f2587da3b6f5c4915d49b0ee39/json_stream_generator/__init__.py#L19). If you mean the Django / DRF response objects, I've just added a bit that makes the construction of these objects more explicit (and I hope more correct with respect to the goals).

Nope, I meant with any serializer in `serializers.py` it seems they are returning only a `list` of features. Isn't it expected to return a `dict` having one of the values a list of features?

Reply:
Since the unit tests succeed, since our custom renderers require a dict as first arg, and since this arg = the value of `serializer.data`, where `serializer` is any instance of a Serializer, I would put my money on the idea that the answer is: a `dict` (probably an OrderedDict, seeing from rest_framework's code).

Reply:
```
# IMO good
(django-ogcapif) *[72b7284][~/work/opengisch/django-ogcapif]$ curl http://0.0.0.0:7180/oapif/collections/tests.point_2056_10fields/items | jq | head 
{
  "links": [
    {
      "type": "application/geo+json",
      "rel": "next",
      "title": "items (next)",
      "href": "http://0.0.0.0:7180/oapif/collections/tests.point_2056_10fields/items?limit=1000&offset=1000"
    },
    {
      "type": "application/geo+json",

# IMO good
(django-ogcapif) *[72b7284][~/work/opengisch/django-ogcapif]$ curl http://0.0.0.0:7180/oapif/collections/tests.point_2056_10fields/items\?format\=json\&encoder\=ujson | jq | head
{
  "links": [
    {
      "type": "application/geo+json",
      "rel": "next",
      "title": "items (next)",
      "href": "http://0.0.0.0:7180/oapif/collections/tests.point_2056_10fields/items?encoder=ujson&format=json&limit=1000&offset=1000"
    },
    {
      "type": "application/geo+json",

# IMO not good
(django-ogcapif) *[72b7284][~/work/opengisch/django-ogcapif]$ curl http://0.0.0.0:7180/oapif/collections/tests.point_2056_10fields/items\?format\=json\&streaming\=true | jq | head 
[
  {
    "id": "30ea6bb0-467f-4f26-9f11-c096fbd70db5",
    "type": "Feature",
    "geometry": {
      "bbox": [
        2508500,
        1152000,
        2508500,
        1152000
```

I believe the third call is a different JSON structure, it is part of the specs, or I am missing something?

Reply:
It's a good point. How would you configure a StreamingHttpResponse in such a way that it produces a dict? Iterators by themselves cannot produce dicts so I wonder.

Reply:
If this is problematic, there are two things to be done about unit tests:
1) we need to fix the unit tests so they cover this scenario, as returning an array is not part of the OpenAPI specification AFAIK.
2) we should not entirely rely on unit tests. Checking the produced endpoints output manually is never a bad idea.

> It's a good point. How would you configure a StreamingHttpResponse in such a way that it produces a dict?
I don't think the problem is in `StreamingHttpResponse`, as it just serves bytes, whenever they are parts of an image, csv rows or part of JSON document.

> Iterators by themselves cannot produce dicts so I wonder.
Also IMO there is no specific limitation with iterators, it is more that `JSONStreamingRenderer` might only support lists for encoding, didn't check their code yet. If you need help, we can have a look at the `JSONStreamingRenderer` library code together.
For example in my (very slow) branch I have the stream to spit key/value pairs. https://github.com/opengisch/django-ogcapif/compare/main...streaming

Reply:
Okay, I know see what you meant, my bad, I was looking at the wrong thing. Yes, none of the renderers provided in this PR wraps the results in this "metadata" dict that you have seen returned by the standard renderers. Instead they just return the results as a list. I must admit I didn't pay much attention to conformance because Denis was mostly interested in timing the request-response round-trip. Or so I understood. If there is more for me to do regarding conformance, happy to do so, but I am not sure what the fate of the prototype is (I am not sure if we are still using it as a test bench or if we want to make everything clean and become a reference implementation).

But if we want to make things clean and become a reference implementation, we probably need to talk about conformance, not just about "toy" renderers like these, but about the entire JSON API. For example, if I look at https://demo.pygeoapi.io/master/collections/lakes/items?f=json, I don't see a wild difference with the return values of the these toy renderers.